Add VXLAN Tunnel Endpoint (VTEP) Integration (#17370)
cilium-team/cilium: Documentation/gettingstarted/vtep.rst

   and running. This guide also assumes VTEP devices has been configured with
   VTEP endpoint IP, VTEP CIDRs, VTEP MAC addresses (VTEP MAC). The VXLAN network
   identifier (VNI) *must* be configured as VNI ``2``, which represents traffic
   from the VTEP as the world identity. See :ref:`reserved_labels` for more details.

Enable VXLAN Tunnel Endpoint (VTEP) support
===========================================

This feature requires a Linux 5.4 kernel, and is disabled by default. To enable the
feature, you must explicitly enable the feature and also specify the IPs, CIDR ranges
and MACs for each VTEP device as part of the configuration.

.. tabs::

    .. group-tab:: Helm

        If you installed Cilium via ``helm install``, you may enable
        the VTEP support with the following command:

        .. parsed-literal::

           helm upgrade cilium |CHART_RELEASE| \\
              --namespace kube-system \\
              --reuse-values \\
              --set vtep.enabled  = true \\
              --set vtep.endpoint = "endpoint-1-IP    endpoint-2-IP" \\
              --set vtep.CIDR =     "endpoint-1-CIDR  endpoint-2-CIDR" \\

>               --set vtep.MAC  =     "endpoint-1-MAC   endpoint-2-MAC" \\

Reviewer: Have you tried rendering this page and running the commands how they appear?

I'm worried that `--set vtep.MAC = "..."` will set `vtep.MAC` to `=` rather than the actual target string, or even cause Helm to throw an error.

What I'd expect is either the setting+`=`+value are all together, eg `--set vtep.MAC="..."` OR maybe there is no equals and you can align the options nicely like this, eg `--set vtep.MAC "..."`.

Author: I didn't try helm, not sure how to try it when this PR is not in official  cilium helm repo, maybe there is way to try it?

Reviewer: Yes, you can use helm directly against the local filesystem, ie `helm install cilium ./install/kubernetes/cilium ...`.

Author: thanks, I tested, it looks there should be `no space` between `vtep.*` and `=` and `" ... "`,  no `=` does not work, 
 below works
```
helm template cilium  ./install/kubernetes/cilium  \
--namespace kube-system \
--set vtep.enabled="true" \
--set vtep.Endpoint="endpoint-1-IP    endpoint-2-IP" \
--set vtep.CIDR="endpoint-1-CIDR  endpoint-2-CIDR" \
--set vtep.MAC="endpoint-1-MAC   endpoint-2-MAC" > test.yaml
```
result in test.yaml
```
  enable-vtep: "true"
  vtep-endpoint: "endpoint-1-IP    endpoint-2-IP"
  vtep-cidr: "endpoint-1-CIDR  endpoint-2-CIDR"
  vtep-mac: "endpoint-1-MAC   endpoint-2-MAC"
```

if you have no objections, i will update the vtep.rst accordingly.

Reviewer: :+1: thanks for testing it out, makes sense, please do update these to remove the spaces around the `=`.

Author: yes I updated, done

    |   lxcxxx                 |
    |      |                   |
    +------+-----cilium_vxlan--+

.. code-block:: shell-session

   # # Create a vxlan device and set the MAC address.
   # ip link add vxlan2 type vxlan id 2 dstport 8472 local 10.169.72.236 dev ens192
   # ip link set dev vxlan2 address 82:36:4c:98:2e:56
   # ip link set vxlan2 up

   # # Configure the VTEP with IP 10.1.1.236 to handle CIDR 10.1.1.0/24.
   # ip addr add 10.1.1.236/24 dev vxlan2

   # # Assume Cilium podCIDR network is 10.0.0.0/16, add route to 10.0.0.0/16
   # ip route add 10.0.0.0/16 dev vxlan2  proto kernel  scope link  src 10.1.1.236

   # # Allow Linux VM send ARP broadcast request to Cilium node for POD ARP resolution
   # # through vxlan2 device, note this depend on if Cilium is able to proxy POD ARP
   # # see https://github.com/cilium/cilium/issues/16890
   # bridge fdb append 00:00:00:00:00:00 dst 10.169.72.233 dev vxlan2

If you are managing multiple VTEPs, follow the above process for each instance.
Once the VTEPs are configured, you can configure Cilium to use the MAC, IP and CIDR ranges that
you have configured on the VTEPs. Follow the instructions to :ref:`enable_vtep`.

> To test the VTEP network connectivity

Reviewer: ```suggestion
To test the VTEP network connectivity:
```

Author: done

.. code-block:: shell-session

   # # ping Cilium-managed busybox pod IP 10.0.1.1 for example from Linux VM

>    # ping 10.0.1.1

Reviewer: nit: I think that if we include the successful output of ping (even just a few lines), that will give users a better understanding of what it looks like to successfully finish the guide.

Author: @joestringer to see a successful ping test reply, it also requires manually add pod ARP in the Linux VM. what do you think if I add https://github.com/cilium/cilium/issues/16890 for ARP in this PR?  it will be complete solution and no manual ARP management.

Reviewer: I'm OK to either just add the instructions for the manual step or fold the extra code in. I haven't thought much about the implications of the ARP responder yet so I'm less certain about what additional changes may be required on that side.

Author: yes, that is why I did not put the ARP in this PR

        .. parsed-literal::

           helm upgrade cilium |CHART_RELEASE| \
              --namespace kube-system \
              --reuse-values \
              --set vtep.enabled="true" \
              --set vtep.Endpoint="endpoint-1-IP endpoint-2-IP" \
              --set vtep.CIDR="endpoint-1-CIDR   endpoint-2-CIDR" \
              --set vtep.MAC="endpoint-1-MAC     endpoint-2-MAC" \

    .. group-tab:: ConfigMap

       VTEP support can be enabled by setting the
       following options in the ``cilium-config`` ConfigMap:

       .. code-block:: yaml

          enable-vtep:   "true"
          vtep-endpoint: "endpoint-1-IP    endpoint-2-IP"
          vtep-cidr:     "endpoint-1-CIDR  endpoint-2-CIDR"
          vtep-mac:      "endpoint-1-MAC   endpoint-2-MAC"

       Restart Cilium daemonset:

       .. code-block:: shell-session

          kubectl -n $CILIUM_NAMESPACE rollout restart ds/cilium

How to test VXLAN Tunnel Endpoint (VTEP) support
================================================

Start up a Linux VM with node network connectivity to Cilium node.
To configure the Linux VM, you will need to be ``root`` user or
run the commands below using ``sudo``.

> ```

Reviewer: For these files, I think this needs to be `::`.

I forget why I thought that you might need the ```` ``` ````, but definitely here it's not necessary. Looking at how the `test/vtep/README.rst` file is formatted on github, I think you need `::` there as well. Sorry to mislead you, I must have just been mistaken.

Author: no worry, I think both should be fixed now with `::` and `space line` before the diagram title

   # ip link set dev vxlan2 address 82:36:4c:98:2e:56
   # ip link set vxlan2 up

   # # Configure the VTEP with IP 10.1.1.236 to handle CIDR 10.1.1.0/24.
   # ip addr add 10.1.1.236/24 dev vxlan2

   # # Assume Cilium podCIDR network is 10.0.0.0/16, add route to 10.0.0.0/16
   # ip route add 10.0.0.0/16 dev vxlan2  proto kernel  scope link  src 10.1.1.236

   # # Allow Linux VM send ARP broadcast request to Cilium node for POD ARP resolution
   # # through vxlan2 device, note this depend on if Cilium is able to proxy POD ARP
   # # see https://github.com/cilium/cilium/issues/16890
   # bridge fdb append 00:00:00:00:00:00 dst 10.169.72.233 dev vxlan2

If you are managing multiple VTEPs, follow the above process for each instance.
Once the VTEPs are configured, you can configure Cilium to use the MAC, IP and CIDR ranges that
you have configured on the VTEPs. Follow the instructions to :ref:`enable_vtep`.

To test the VTEP network connectivity:

.. code-block:: shell-session

   # # ping Cilium-managed busybox pod IP 10.0.1.1 for example from Linux VM
   # ping 10.0.1.1

Limitations
===========

* This feature does not work with `CiliumClusterwideNetworkPolicy`.

> * The VTEP IPs are associated directly with the "World" identity (see :ref:`reserved_labels`).  This means that network policy that allows ``toCIDR`` / ``fromCIDR`` ranges will not work with VTEP IPs. Please file an issue if you have a use case where pods may want to communicate directly with VTEPs.

Reviewer: Given the discussion I just started on one of the [recent threads](https://github.com/cilium/cilium/pull/17370#discussion_r728536725), I think we can actually be more broad about the incompatibility with CIDR policy:
```suggestion
* This feature does not work with ``ToCIDR``, ``FromCIDR`` or ``ToFQDN`` policies.
```

Author: done

***********************************************
VXLAN Tunnel Endpoint (VTEP) Integration (beta)
***********************************************

.. include:: ../beta.rst

The VTEP integration allows third party VTEP devices to send and receive traffic to
and from Cilium-managed pods directly using VXLAN. This allows for example external
load balancers like BIG-IP to load balance traffic to Cilium-managed pods using VXLAN.

This document explains how to enable VTEP support and configure Cilium with VTEP
endpoint IPs, CIDRs, and MAC addresses.

.. note::

   This guide assumes that Cilium has been correctly installed in your
   Kubernetes cluster. Please see :ref:`k8s_quick_install` for more
   information. If unsure, run ``cilium status`` and validate that Cilium is up
   and running. This guide also assumes VTEP devices has been configured with
   VTEP endpoint IP, VTEP CIDRs, VTEP MAC addresses (VTEP MAC). The VXLAN network
   identifier (VNI) *must* be configured as VNI ``2``, which represents traffic
   from the VTEP as the world identity. See :ref:`reserved_labels` for more details.

Enable VXLAN Tunnel Endpoint (VTEP) support
===========================================

> This feature requires a Linux 5.2 kernel, and is disabled by default. When enabling the

Reviewer: Typically we use the "X.Y or later" wording to denote which upstream Linux is required.

```suggestion
This feature requires a Linux 5.2 kernel or later, and is disabled by default. When enabling the
```

Author: done

VTEP integration, you must also specify the IPs, CIDR ranges and MACs for each VTEP device
as part of the configuration.

> 

Reviewer: If the feature requires policy to be disabled, I think it makes sense to have a warning something like:
```suggestion
.. warning::

   This feature is in beta, and is currently incompatible with network policy. The instructions below will specify to disable network policy in order to enable the feature for getting started. This restriction will be lifted when the feature graduates from beta. This work is tracked in :gh-issue:`12345`.
```

(Except you'll want to file a new issue something like "Enable policy for VTEP feature" and put the number for the new issue into the text above).

Then, for the guide to work, the install instructions would need to explicitly disable policy.

Author: done

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _enable_vtep:

*******************************************

> VXLAN Tunnel Endpoint (VTEP) support (beta)

Reviewer 2: I'd drop `support` from the title here, to be consistent with other entries.
```suggestion
VXLAN Tunnel Endpoint (beta)
```

Reviewer: What's a bit odd with this one, which is different from other features, is that we're not implementing VTEP. We're implementing an integration feature for VTEP devices. Maybe `VXLAN Tunnel Endpoint (VTEP) integration (beta)` would be a better alternative?

Author: yes better, got it

Author: done

*******************************************

.. include:: ../beta.rst

The VTEP support allows third party VTEP devices to send or receive traffic to
or from Cilium-managed pod directly using VXLAN. This allows for example external

> load balancers like BIG-IP to load balance traffic to Cilium-managed pod using VXLAN

Reviewer 2: ```suggestion
The VTEP support allows third-party VTEP devices to send and receive traffic to
and from Cilium-managed pods directly using VXLAN. This allows for example external
load balancers like BIG-IP to load balance traffic to Cilium-managed pods using VXLAN.
```

Author: done

This document explains how to enable VTEP support and configure Cilium with VTEP
endpoint IPs, CIDRs, and MAC addresses.

.. note::

   This guide assumes that Cilium has been correctly installed in your
   Kubernetes cluster. Please see :ref:`k8s_quick_install` for more
   information. If unsure, run ``cilium status`` and validate that Cilium is up
   and running. This guide also assumes VTEP devices has been configured with
   VTEP endpoint IP, VTEP CIDRs, VTEP MAC addresses (VTEP MAC). The VXLAN network
   identifier (VNI) *must* be configured as VNI ``2``, which represents traffic
   from the VTEP as the world identity. See :ref:`reserved_labels` for more details.

Enable VXLAN Tunnel Endpoint (VTEP) support
===========================================

> This feature requires a Linux 5.4 kernel, and is disabled by default. To enable the

Reviewer 2: ```suggestion
This feature requires a Linux 5.2 kernel, and is disabled by default. To enable the
```

Author: 5.2-5.3 kernel have issues reported in https://github.com/cilium/cilium/issues/17528

Author: done

> feature, you must explicitly enable the feature and also specify the IPs, CIDR ranges

Reviewer 2: ```suggestion
This feature requires a Linux 5.4 kernel, and is disabled by default. When enabling
the VTEP support, you must also specify the IPs, CIDR ranges
```

Author: done

       Restart Cilium daemonset:

       .. code-block:: shell-session

          kubectl -n $CILIUM_NAMESPACE rollout restart ds/cilium

How to test VXLAN Tunnel Endpoint (VTEP) support
================================================

Start up a Linux VM with node network connectivity to Cilium node.
To configure the Linux VM, you will need to be ``root`` user or
run the commands below using ``sudo``.

::

     Test VTEP support

     Node IP: 10.169.72.233
    +--------------------------+            VM IP: 10.169.72.236
    |                          |            +------------------+
    | CiliumNode               |            |  Linux VM        |
    |                          |            |                  |
    |  +---------+             |            |                  |
    |  | busybox |             |            |                  |
    |  |         |           ens192<------>ens192              |
    |  +--eth0---+             |            |                  |
    |      |                   |            +-----vxlan2-------+
    |      |                   |
    |   lxcxxx                 |
    |      |                   |
    +------+-----cilium_vxlan--+

> .. code-block:: shell-session

Reviewer 2: If you use `shell` instead of `shell-session` here, you won't need to prepend `#` to each line.

Reviewer: Really? I don't see any cases in the tree already that uses `.. code-block:: shell` cc @qmonnet

Reviewer 3: I suppose Paul meant `bash`. We do have some occurrences of `.. code-block:: bash`.

Reviewer 2: Both should work as I believe the highlighter is Pygments. Let's go with `bash` for consistency.

Author: I guess I can still keep the one `#` for the command comment, remove all `#` for the bash commands ? I will try it and then maybe "view file" after I push

Author: done

   # # Create a vxlan device and set the MAC address.
   # ip link add vxlan2 type vxlan id 2 dstport 8472 local 10.169.72.236 dev ens192
   # ip link set dev vxlan2 address 82:36:4c:98:2e:56
   # ip link set vxlan2 up

   # # Configure the VTEP with IP 10.1.1.236 to handle CIDR 10.1.1.0/24.
   # ip addr add 10.1.1.236/24 dev vxlan2

   # # Assume Cilium podCIDR network is 10.0.0.0/16, add route to 10.0.0.0/16
   # ip route add 10.0.0.0/16 dev vxlan2  proto kernel  scope link  src 10.1.1.236

   # # Allow Linux VM send ARP broadcast request to Cilium node for busybox POD ARP resolution

>    # # through vxlan2 device, note this depend on if Cilium is able to proxy busybox POD ARP

Reviewer 2: Nit:
```suggestion
   # # Allow Linux VM send ARP broadcast request to Cilium node for busybox pod ARP resolution
   # # through vxlan2 device, note this depend on if Cilium is able to proxy busybox pod ARP
```

Author: done

   # # see https://github.com/cilium/cilium/issues/16890
   # bridge fdb append 00:00:00:00:00:00 dst 10.169.72.233 dev vxlan2

   # # Another way is to manually add busybox POD ARP address with Cilium node 10.169.72.233
   # # cilium_vxlan interface MAC address like below
   # bridge fdb append  <cilium_vxlan MAC> dst 10.169.72.233 dev vxlan2
   # arp -i vxlan2 -s <busybox POD IP> <cilium_vxlan MAC>

If you are managing multiple VTEPs, follow the above process for each instance.
Once the VTEPs are configured, you can configure Cilium to use the MAC, IP and CIDR ranges that
you have configured on the VTEPs. Follow the instructions to :ref:`enable_vtep`.

To test the VTEP network connectivity:

> .. code-block:: shell-session

Reviewer 2: Ditto here regarding `shell`.

Author: done

   # # ping Cilium-managed busybox pod IP 10.0.1.1 for example from Linux VM
   # ping 10.0.1.1

Limitations
===========

> * This feature does not work with `CiliumClusterwideNetworkPolicy`.

Reviewer 2: Why is that?

Author: I think @joestringer suggested to add that :)

Reviewer: Looking back over, I see that there's currently only an implementation in `bpf/bpf_lxc.c`, so in general it will not apply to host traffic (which is handled by `bpf/bpf_host.c`). Beyond that I can't think of anything specific to CCNP, it's just the general network policy concerns.

From that perspective, it's probably more accurate to say:

```suggestion
* This feature does not work with traffic from the host network namespace (including pods with `hostNetwork=true`).
```

For reference, the previous related discussion was [here](https://github.com/cilium/cilium/pull/17370#discussion_r727394642) but unfortunately past me didn't think to write anything more insightful than `One area where this becomes even more curious is when you consider ClusterwideNetworkPolicy and the way that individual nodes may need to send traffic towards the VTEPs.`

Author: yes indeed pods with `hostNetwork=true` does not work with this feature as I tested.

Author: done

   ip addr add 10.1.1.236/24 dev vxlan2
   # Assume Cilium podCIDR network is 10.0.0.0/16, add route to 10.0.0.0/16
   ip route add 10.0.0.0/16 dev vxlan2  proto kernel  scope link  src 10.1.1.236
   # Allow Linux VM send ARP broadcast request to Cilium node for busybox pod ARP resolution
   # through vxlan2 device, note this depend on if Cilium is able to proxy busybox pod ARP
   # see https://github.com/cilium/cilium/issues/16890
   bridge fdb append 00:00:00:00:00:00 dst 10.169.72.233 dev vxlan2
   # Another way is to manually add busybox pod ARP address with Cilium node 10.169.72.233
   # cilium_vxlan interface MAC address like below
   bridge fdb append  <cilium_vxlan MAC> dst 10.169.72.233 dev vxlan2
   arp -i vxlan2 -s <busybox pod IP> <cilium_vxlan MAC>

If you are managing multiple VTEPs, follow the above process for each instance.
Once the VTEPs are configured, you can configure Cilium to use the MAC, IP and CIDR ranges that
you have configured on the VTEPs. Follow the instructions to :ref:`enable_vtep`.

To test the VTEP network connectivity:

.. code-block:: bash

   # ping Cilium-managed busybox pod IP 10.0.1.1 for example from Linux VM
   ping 10.0.1.1

> Limitations

Reviewer: Maybe one more, this doesn't work with ipsec mode? Probably worth mentioning since this PR is adding dedicated code that will be incompatble with ipsec mode. Even if we do properly support tunnel + ipsec, this feature would need additional validation.

Author: this PR does not work with ipsec for VTEP over the tunnel, but it should not interfere with ipsec for non-VTEP connectivity over tunnel, yes we can add notes mentioning ipsec is not supported for this PR for VTEP integration
